libs/cardano-ledger-core/src/Cardano/Ledger/UMap.hs

```diff
@@ -129,138 +152,242 @@ instance DecCBOR RDPair where
+  Set Ptr ->
+  StrictMaybe (KeyHash 'StakePool c) ->
+  StrictMaybe (Credential 'Voting c) ->
+  UElem c
+pattern UElemP i j k l <- (uelemAsTuple -> (i, j, k, l))
   where
-    Triple a b c =
-      case (a, b, c) of
-        (SNothing, SI.Tip, SNothing) -> TEEE
-        (SNothing, SI.Tip, SJust x) -> TEEF x
-        (SNothing, x, SNothing) -> TEFE x
-        (SNothing, x, SJust y) -> TEFF x y
-        (SJust x, SI.Tip, SNothing) -> TFEE x
-        (SJust x, SI.Tip, SJust y) -> TFEF x y
-        (SJust x, y, SNothing) -> TFFE x y
-        (SJust x, y, SJust z) -> TFFF x y z
-
-{-# COMPLETE Triple #-}
-
-instance Show (Trip c) where
-  show (Triple a b c) = "(Triple (" ++ show a ++ ", " ++ show b ++ ", " ++ show c ++ "))"
-
--- =====================================================
-
--- | A unified map represents 4 Maps with domain @(Credential 'Staking c)@ for
---   keys and one more in the inverse direction with @Ptr@ for keys and @(Credential 'Staking c)@ for values.
-data UMap c = UMap !(Map (Credential 'Staking c) (Trip c)) !(Map Ptr (Credential 'Staking c))
-  deriving (Eq, Generic, NoThunks, NFData)
-
-instance Show (UMap c) where
-  show (UMap tripmap ptrmap) =
+    UElemP i j k l = case (i, j, k, l) of
+      (SNothing, SI.Tip, SNothing, SNothing) -> TEEEE
+      (SNothing, SI.Tip, SNothing, SJust d) -> TEEEF d
+      (SNothing, SI.Tip, SJust s, SNothing) -> TEEFE s
+      (SNothing, SI.Tip, SJust s, SJust d) -> TEEFF s d
+      (SNothing, p, SNothing, SNothing) -> TEFEE p
+      (SNothing, p, SNothing, SJust d) -> TEFEF p d
+      (SNothing, p, SJust s, SNothing) -> TEFFE p s
+      (SNothing, p, SJust s, SJust d) -> TEFFF p s d
+      (SJust r, SI.Tip, SNothing, SNothing) -> TFEEE r
+      (SJust r, SI.Tip, SNothing, SJust d) -> TFEEF r d
+      (SJust r, SI.Tip, SJust s, SNothing) -> TFEFE r s
+      (SJust r, SI.Tip, SJust s, SJust d) -> TFEFF r s d
+      (SJust r, p, SNothing, SNothing) -> TFFEE r p
+      (SJust r, p, SNothing, SJust d) -> TFFEF r p d
+      (SJust r, p, SJust s, SNothing) -> TFFFE r p s
+      (SJust r, p, SJust s, SJust d) -> TFFFF r p s d
+
+{-# COMPLETE UElemP #-}
+
+instance Show (UElem c) where
+  show (UElemP a b c d) =
     unlines
-      [ "(UMap ("
-      , show tripmap
-      , show ptrmap
+      [ "(UElemP ("
+      , show a <> ", "
+      , show b <> ", "
+      , show c <> ", "
+      , show d
       , "))"
       ]
 
+-- | A unified map represents 4 Maps with domain @(Credential 'Staking c)@
+--
+-- 1) Map (Credential 'Staking c) RDPair  -- (RDPair rewardCoin depositCoin)
+-- 2) Map (Credential 'Staking c) (Set Ptr)
+-- 3) Map (Credential 'Staking c) (StrictMaybe (KeyHash 'StakePool c))
+-- 4) Map (Credential 'Staking c) (StrictMaybe (Credential 'Voting c))
+-- and one more map in the inverse direction with @Ptr@ for keys and @(Credential 'Staking c)@ for values.
+data UMap c = UMap
+  { umElem :: !(Map (Credential 'Staking c) (UElem c))
+  , umPtr :: !(Map Ptr (Credential 'Staking c))
```

Comment:
Something to consider, not too important though:
```suggestion
  { umElems :: !(Map (Credential 'Staking c) (UElem c))
  , umPtrs :: !(Map Ptr (Credential 'Staking c))
```

Reply:
Defintely!